Define new attestation mechanisms

Kakemone opened the pull request:
Added CertifyAndQuote mechanism to enable key and platform attestation.

Kakemone committed:
Define new attestation mechanisms

Added CertifyAndQuote mechanism to enable key and platform attestation.

Kakemone committed:
Define new attestation mechanisms 

Added CertifyAndQuote mechanism to enable key and platform attestation.

ionut-arm commented on src/operations/attest_key.rs:
```diff
@@ -25,6 +25,16 @@ pub enum Operation {
+    CertifyAndQuote {
+        /// Name of key to be attested
+        attested_key_name: String,
+        /// The nonce to be used in the TLS handshake
+        #[derivative(Debug = "ignore")]
+        nonce: Zeroizing<Vec<u8>>,
```
The nonce isn't actually secret or sensitive, so you can just leave it as `Vec<u8>` here.

ionut-arm commented on src/operations_protobuf/convert_attest_key.rs:
```diff
@@ -129,6 +182,32 @@ mod test {
         let op_secret = vec![0x11; 16];
+        let op_nonce = vec![0x88; 8];
+
+        let proto = OperationProto {
```
I would probably put these checks into separate tests, just to make it easier to debug/figure out what's happening if one of them fails. You can probably rename the existing ones something to do with "activate credential"

Kakemone replied:
How do names like `attest_key_op_from_proto_CertifyAndQuote` and `attest_key_op_from_proto_ActivateCredential` sound? Or did you have a different idea?

ionut-arm replied:
That sounds alright!

ionut-arm commented on src/operations_protobuf/convert_attest_key.rs:
```diff
@@ -191,6 +289,33 @@ mod test {
+            assert_eq!(attested_key_name, op_attested_key_name);
+            assert_eq!(attesting_key_name, op_attesting_key_name);
+            assert_eq!(nonce, op_nonce);
+        }
```
I think we should have an `else` here, in case the conversion produced the wrong thing - if you don't actually get what you expect in that `if let`, this branch is never taken and the test currently passes, but it should actually fail

ionut-arm replied:
Oh, and you should also fix the existing tests for this! (just the ones in this file, don't know if other files are affected)

Kakemone replied:
I've done this but left `panic!("'if let' branch not taken")` in the else statement. This response probably needs to be better but it should be a reasonable placeholder.

ionut-arm reviewed (commented):
Thanks, the code looks good, I had mostly "structural" comments

Kakemone committed:
Define new attestation mechanisms

Added CertifyAndQuote mechanism to enable key and platform attestation.

Kakemone committed:
Define new attestation mechanisms

Added CertifyAndQuote mechanism to enable key and platform attestation.

ionut-arm commented:
By the way, it's probably worth adjusting commit title/description to what is in the commit, not what the bigger change set is intended to do. Don't worry about the existing commits, though, just for future reference

Kakemone committed:
Add CertifyAndQuote

Added the structures required for the CertifyAndQuote mechanism

ionut-arm commented:
One thing I forgot about is the zeroizing for protobuf values, have a look [here](https://github.com/ionut-arm/parsec-interface-rs/blob/attested-tls/src/operations_protobuf/generated_ops/mod.rs#L345). You can probably change the `if let` to a `match` and zeroize the attestation blobs.

Kakemone committed:
Fixed formatting

Fixed the formatting using `cargo fmt` to get it to pass CI.

Kakemone committed:
Update implementation of clear_message() for attest_key::Result

Add to implementation of clear_message() so it can handle the result of the CertifyAndQuote mechanism.

Kakemone committed:
Fixed clippy warnings

Signed-off-by: Sam Davis <Sam.Davis@arm.com>

ionut-arm reviewed (approved):
👍🏻
